gadget/update.go

```diff
@@ -124,6 +124,102 @@ type ContentUpdateObserver interface {
 	Canceled() error
 }
 
+func searchForVolumeWithTraits(laidOutVol *LaidOutVolume, traits DiskVolumeDeviceTraits, validateOpts *DiskVolumeValidationOptions) (disks.Disk, error) {
+	if validateOpts == nil {
+		validateOpts = &DiskVolumeValidationOptions{}
+	}
+
+	vol := laidOutVol.Volume
+	// iterate over the different traits, validating whether the resulting disk
+	// actually exists and matches the volume we have in the gadget.yaml
+
+	checkCandidateDisk := func(candidate disks.Disk, method string, providedErr error) bool {
+		if providedErr != nil {
+			if candidate != nil {
+				logger.Debugf("candidate disk %s not appropriate for volume %s because err: %v", candidate.KernelDeviceNode(), vol.Name, providedErr)
+				return false
+			}
+			logger.Debugf("cannot locate disk for volume %s with method %s because err: %v", vol.Name, method, providedErr)
+
+			return false
+		}
+		diskLayout, onDiskErr := OnDiskVolumeFromDevice(candidate.KernelDeviceNode())
+		if onDiskErr != nil {
+			// unexpected in reality, we already called one of
+			// DiskFromDeviceName or DiskFromDevicePath to get this reference,
+			// so it's unclear how those methods could return a disk that
+			// OnDiskVolumeFromDevice is unhappy about
+			logger.Debugf("cannot find on disk volume from candidate disk %s: %v", candidate.KernelDeviceNode(), onDiskErr)
+			return false
+		}
+		// then try to validate it by laying out the volume
+		opts := &EnsureLayoutCompatibilityOptions{
+			AssumeCreatablePartitionsCreated: true,
+			AllowImplicitSystemData:          validateOpts.AllowImplicitSystemData,
+			ExpectedStructureEncryption:      validateOpts.ExpectedStructureEncryption,
+		}
+		ensureErr := EnsureLayoutCompatibility(laidOutVol, diskLayout, opts)
+		if ensureErr != nil {
+			logger.Debugf("candidate disk %s not appropriate for volume %s due to incompatibility: %v", candidate.KernelDeviceNode(), vol.Name, ensureErr)
+			return false
+		}
+
+		// success, we found it
+		return true
+	}
+
+	// first try the kernel device path if it is set
+	if traits.OriginalDevicePath != "" {
+		disk, err := disks.DiskFromDevicePath(traits.OriginalDevicePath)
+		if ok := checkCandidateDisk(disk, "device path", err); ok {
```

Comment:
the probably returned error at some point?
```suggestion
		if checkCandidateDisk(disk, "device path", err) {
```

Reply:
yes good catch